apps/editing-toolkit/editing-toolkit-plugin/tutorials/class-wpcom-tutorials-endpoint.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * Tutorials endpoint (Think: "Guided Tours").
+ *
+ * @package A8C\FSE
+ */
+
+/**
+ * WPCom_Tutorials_Controller class
+ */
+class WPCom_Tutorials_Endpoint extends WP_REST_Controller {
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		$this->namespace                       = 'wpcom/v2';
+		$this->rest_base                       = 'tutorials';
+		$this->wpcom_is_site_specific_endpoint = false;
+	}
+
+	/**
+	 * Callback for registering our REST routes
+	 */
+	public function register_rest_routes() {
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_tutorials' ),
+					'permission_callback' => 'is_user_logged_in',
+				),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/(?P<tutorial_id>[\w-]+)',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_tutorial' ),
+					'permission_callback' => 'is_user_logged_in',
+				),
+				array(
+					'methods'             => WP_REST_Server::EDITABLE,
+					'callback'            => array( $this, 'update_tutorial' ),
+					'permission_callback' => 'is_user_logged_in',
+					'args'                => $this->get_collection_params(),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Callback for the `/tutorials` endpoint
+	 *
+	 * @return WP_REST_Response
+	 */
+	public function get_tutorials() {
+		$tutorials = wpcom_get_registered_tutorial_ids();
```

Comment:
I might be totally mistaken, but don't we want to loop through all ids and retrieve tutorial data to return from this endpoint?

From what I can tell, this endpoint only returns a set of tutorial ids.

Reply:
Good question in the absence of my rationale, which never made it out of a mental post-it note.

If we knew we were the only ones who would ever use this API, I would be in agreement. We would simply query the `/tutorials` endpoint and immediately have access to all of the data we need to get up and running.

But if we get this right, Tutorials/Tours will be seen as very useful functionality and use should expand across our products. This will at some point make the full-freight `/tutorials` endpoint expensive and unwieldy, returning much more data than is required, and spending more time processing it, making the endpoint slower over time.

And so my own thinking was that, in practice, we would probably not be hitting the `/tutorials` endpoint much, if at all, but rather hardcoding the tutorial IDs into our frontend code and querying the individually-needed endpoints as needed. Yes, this increases the number of total API requests in areas of the product that use > 1 Tutorial, but this appeared to be an ok tradeoff compared with bloating the `/tutorials` endpoint. (We also have the capability to batch multiple requests into one, at least in theory.)

Now, perhaps this is premature optimization, but overall I just think it's more sane to fetch the required data rather than all of it, knowing that "all of it" should grow over time.

Reply:
This makes total sense. I am, however, wondering whether or not we could use some query params to filter our data? 

I imagine at a high level we could categorize groups of tutorials together ( `editor`, `woocommerce`, `jetpack` ). Then, we could subsequently use the group keys as query params to pull the correct data.

It feels like this would be easier to reason about and debug than to use hard coded ids

Curious about your thoughts 🤔

Reply:
Batching a bunch of tutorial IDs sounds potentially useful, but it might be better in terms of frontend code reuse/encapsulation to have something like a `Tutorials` component that includes automatic fetching of the stuff from the `/tutorials/editor-basics` endpoint in the following instance:

`<Tutorial id="editor-basics" steps={} /*etc*/ />`

This kind of encapsulation makes sense to me, and if we need a performance gain, we control the server side and client side and could definitely figure out a way to fetch everything in a single request and cache it properly. Let's do that work if it needs to happen but it feels like premature optimization to me.

Reply:
I think im in agreement with the encapsulation approach @mattwiebe is describing.

In general I think we would use `/tutorials/$tutorial_id` much more often than `/tutorials` from the front-end. Im not thinking of much use for `tutorials` itself past debugging and validating that a tutorial is actually registered. 

I would imagine the front end we would define a list of tutorials  (IDs, tasks, steps, dependent selectors, required blocks for those steps, etc. etc. ), have some logic to determine which need to be surfaced in 'Next Steps' at any particular time/situation, pass the data for those specific tutorials to encapsulated <TutorialManager> (or whatever we might name them) components, and those components encapsulate all concerns for reading/writing to their respective `tutorials/$tutorial_id` endpoints. 

I don't think @jeyip's ideas are unreasonable and they would enable other forms of managing the tutorials, but initially it seems like the interactions with the api would be a bit more separated (a need to get the data at a higher component level, while the need to write still sits at a lower component level). Not to say there aren't ways around dealing with that. Either way, I think what we currently have should be suitable for our needs, and if we do have the need to batch request full tutorial data by lists of IDs it should be simple enough to add that functionality here or in a new related route.